### PR TITLE
Fix small issues

### PR DESCRIPTION
**Pull request by Alexander-Wilms**
Found most of them using the analysis linked in https://github.com/vcmi/vcmi/issues/3112

### CODE REVIEW

**Comment by Alexander-Wilms** [comment]
You mean for the SDL calls?

Does that matter, since the calling function is written in C++?

**Comment by rilian-la-te** [comment]
Yes, it is matter. If we will not get any bugs on exotic systems - passing nullptr to pure C is also okay.

File: client/NetPacksClient.cpp [hunk 1]
```diff
@@ -367,19 +367,27 @@ void ApplyClientNetPackVisitor::visitGiveBonus(GiveBonus & pack)
...
-
-	CGI->mh->waitForOngoingAnimations();
+	if(CGI)
+	{
+		if(CGI->mh)
+		{
+			CGI->mh->onObjectFadeOut(obj, pack.initiator);
+			CGI->mh->waitForOngoingAnimations();
+		}
+	}
```
Comment (Nordsoft91) [inline client/NetPacksClient.cpp:377]:
```suggestion
	if(CGI && CGI->mh)
	{
		CGI->mh->onObjectFadeOut(obj, pack.initiator);
		CGI->mh->waitForOngoingAnimations();
	}
```

File: client/NetPacksClient.cpp [hunk 2]
```diff
@@ -367,19 +367,27 @@ void ApplyClientNetPackVisitor::visitGiveBonus(GiveBonus & pack)
...
-
-	CGI->mh->waitForOngoingAnimations();
+	if(CGI)
+	{
+		if(CGI->mh)
+		{
+			CGI->mh->onObjectFadeIn(obj, pack.initiator);
+			CGI->mh->waitForOngoingAnimations();
+		}
+	}
```
Comment (Nordsoft91) [inline client/NetPacksClient.cpp:390]:
```suggestion
	if(CGI && CGI->mh)
	{
		CGI->mh->onObjectFadeIn(obj, pack.initiator);
		CGI->mh->waitForOngoingAnimations();
	}
```

File: lib/BasicTypes.cpp [hunk 3]
```diff
@@ -111,7 +111,7 @@ int AFactionMember::moraleValAndBonusList(TConstBonusListPtr & bonusList) const
 	bonusList = getBonusBearer()->getBonuses(moraleSelector, cachingStrMor);
 
 	int32_t maxGoodMorale = VLC->settings()->getVector(EGameSettings::COMBAT_GOOD_MORALE_DICE).size();
-	int32_t maxBadMorale = -VLC->settings()->getVector(EGameSettings::COMBAT_BAD_MORALE_DICE).size();
+	int32_t maxBadMorale = - (int32_t) VLC->settings()->getVector(EGameSettings::COMBAT_BAD_MORALE_DICE).size();
```
Comment (Nordsoft91) [inline lib/BasicTypes.cpp:114]: why is needed?

File: lib/int3.h [hunk 4]
```diff
@@ -82,19 +82,10 @@ class int3
 
 	constexpr bool operator<(const int3 & i) const
 	{
-		if (z < i.z)
+		if (x < i.x || y < i.y || z < i.z)
```
Comment (Nordsoft91) [inline lib/int3.h:85]:
```suggestion
		return z < i.z || y < i.y || x < i.x;
```

File: lib/mapObjects/IMarket.cpp [hunk 5]
```diff
@@ -157,7 +157,7 @@ std::vector<int> IMarket::availableItemsIds(EMarketMode mode) const
 const IMarket * IMarket::castFrom(const CGObjectInstance *obj, bool verbose)
 {
 	auto * imarket = dynamic_cast<const IMarket *>(obj);
-	if(verbose && !imarket)
+	if(verbose && !imarket && obj)
```
Comment (Nordsoft91) [inline lib/mapObjects/IMarket.cpp:160]: if obj is nullptr we simply ignore it?

File: lib/rmg/float3.h [hunk 6]
```diff
@@ -117,20 +117,10 @@ class float3
 
 	bool operator<(const float3 & i) const
 	{
-		if (z<i.z)
+		if (x<i.x || y<i.y || z<i.z)
```
Comment (Nordsoft91) [inline lib/rmg/float3.h:120]:
```suggestion
		return z < i.z || y < i.y || x < i.x;
```

File: server/battles/BattleActionProcessor.cpp [hunk 7]
```diff
@@ -1445,8 +1445,11 @@ bool BattleActionProcessor::makePlayerBattleAction(const CBattleInfoCallback & b
 	else
 	{
 		auto active = battle.battleActiveUnit();
-		if(!active && gameHandler->complain("No active unit in battle!"))
+		if(!active)
+		{
+			gameHandler->complain("No active unit in battle!");
```
Comment (Nordsoft91) [inline server/battles/BattleActionProcessor.cpp:1450]: change is not required here and in other similar places. It's compact style.

File: server/battles/BattleFlowProcessor.cpp [hunk 8]
```diff
@@ -265,6 +265,10 @@ const CStack * BattleFlowProcessor::getNextStack(const CBattleInfoCallback & bat
 		return nullptr;
 
 	auto next = q.front().front();
+
+	if(next == nullptr)
+		return nullptr;
+
```
Comment (Nordsoft91) [inline server/battles/BattleFlowProcessor.cpp:271]:
```suggestion
```

Instead, on line 287 replace
```
if(!next->willMove();
```

with
```
if(!next || !next->willMove())
```

File: server/processors/HeroPoolProcessor.cpp [hunk 9]
```diff
@@ -145,11 +145,17 @@ bool HeroPoolProcessor::hireHero(const ObjectInstanceID & objectID, const HeroTy
...
+		gameHandler->complain("Invalid map object!");
 		return false;
+	}
 
-	if (!playerState && gameHandler->complain("Invalid player!"))
+	if (!playerState)
+	{
+		gameHandler->complain("Invalid player!");
 		return false;
+	}
```
Comment (Nordsoft91) [inline server/processors/HeroPoolProcessor.cpp:158]: unnecessary changes

File: CI/linux-qt6/validate_json.py [hunk 10]
```diff
@@ -12,19 +13,19 @@
...
 
 errors = []
-for path in sorted(Path('.').glob('**/*.json')):
+for path in sorted(Path(".").glob("**/*.json")):
     # because path is an object and not a string
     path_str = str(path)
     try:
-        with open(path_str, 'r') as file:
-            if VALIDATION_TYPE == 'json':
+        with open(path_str, "r") as file:
```
[the remark this answers is not in the document]
Comment (Alexander-Wilms) [reply CI/linux-qt6/validate_json.py:23]: It's the default for [ruff](https://github.com/astral-sh/ruff)

File: lib/mapObjects/IMarket.cpp (hunk 5 above)
Comment (Alexander-Wilms) [reply lib/mapObjects/IMarket.cpp:160]: Oh, right. Should I split the error message in two?

File: lib/BasicTypes.cpp (hunk 3 above)
Comment (Alexander-Wilms) [reply lib/BasicTypes.cpp:114]: https://rules.sonarsource.com/c/RSPEC-876/

File: lib/int3.h [hunk 11]
```diff
@@ -82,19 +82,7 @@ class int3
...
-		if (y < i.y)
-			return true;
-		if (y > i.y)
-			return false;
-		if (x < i.x)
-			return true;
-		if (x > i.x)
-			return false;
-		return false;
+		return z < i.z || y < i.y || x < i.x;
```
Comment (Nordsoft91) [inline lib/int3.h:85]: hm, I'm thinking maybe this is not equivalent to initial implementation which can introduce problems:
`A(0, 0, 1) < B(1, 0, 0)`
Before change: false
After change: true

Comment (Nordsoft91) [reply lib/int3.h]: equivalent condition looks following:
`return z < i.z || (z == i.z && (y < i.y || (y == i.y && x < i.x)));`

File: client/widgets/Buttons.h [hunk 12]
```diff
@@ -67,7 +67,7 @@ class CButton : public CKeyShortcut
 	void setBorderColor(std::optional<ColorRGBA> borderColor);
 
 	/// adds one more callback to on-click actions
-	void addCallback(std::function<void()> callback);
+	void addCallback(std::function<void()> const & callback);
```
Comment (kambala-decapitator) [inline client/widgets/Buttons.h:70]:
```suggestion
	void addCallback(const std::function<void()> & callback);
```

File: client/widgets/Buttons.cpp [hunk 13]
```diff
@@ -66,7 +66,7 @@ void CButton::setBorderColor(std::optional<ColorRGBA> newBorderColor)
 	borderColor = newBorderColor;
 }
 
-void CButton::addCallback(std::function<void()> callback)
+void CButton::addCallback(std::function<void()> const & callback)
```
Comment (kambala-decapitator) [inline client/widgets/Buttons.cpp:69]:
```suggestion
void CButton::addCallback(const std::function<void()> & callback)
```

File: launcher/main.cpp [hunk 14]
```diff
@@ -35,7 +35,7 @@ int main(int argc, char * argv[])
 
 	MainWindow mainWindow;
 	mainWindow.show();
-	result = vcmilauncher.exec();
+	result = QApplication::exec();
```
Comment (kambala-decapitator) [inline launcher/main.cpp:38]: now there'll be warning "variable vcmilauncher is unused"

**Comment by kambala-decapitator** [comment]
> @Alexander-Wilms I think if you request C code from C++ functions, using NULL is better than nullptr. C does not know about nullptr.

nullptr is just a helper so that you can't assign null pointer to an int

File: launcher/main.cpp (hunk 14 above)
Comment (Alexander-Wilms) [reply launcher/main.cpp:38]: Right, will drop the commit.

File: AI/BattleAI/BattleAI.cpp [hunk 15]
```diff
@@ -197,7 +197,7 @@ BattleAction CBattleAI::useCatapult(const BattleID & battleID, const CStack * st
 	}
 	else
 	{
-		EWallPart wallParts[] = {
+		std::array<EWallPart, 7> wallParts {
```
Comment (IvanSavenko) [inline AI/BattleAI/BattleAI.cpp:200]:
```suggestion
		std::array wallParts {
```
(yes, this works)
https://en.cppreference.com/w/cpp/language/class_template_argument_deduction

`std::array<EWallPart, 7>` form is actually somewhat dangerous since it does not checks for initializer size.
If you initialize such array with less than 7 elements it will still compile but won't work as expected
